subsys/usb/host/class/usbh_cdc_ecm.c

```diff
@@ -0,0 +1,1157 @@
+			     struct usbh_cdc_ecm_xfer_params *const param)
+{
+	int ret;
+
+	param->xfer = NULL;
+
+	if (!param || !param->ep_addr || !param->cb || !param->buf) {
+		return -EINVAL;
+	}
+
+	param->xfer = usbh_xfer_alloc(udev, param->ep_addr, param->cb, param->cb_priv);
+	if (!param->xfer) {
+		return -ENOMEM;
+	}
+
+	ret = usbh_xfer_buf_add(udev, param->xfer, param->buf);
+	if (ret) {
+		goto cleanup;
+	}
+
+	ret = usbh_xfer_enqueue(udev, param->xfer);
+	if (ret) {
+		goto cleanup;
+	}
+
+	goto done;
+
+cleanup:
+	(void)usbh_xfer_free(udev, param->xfer);
+
+done:
+	return ret;
+}
+
+static int usbh_cdc_ecm_comm_rx(struct usbh_cdc_ecm_data *const data);
+static int usbh_cdc_ecm_data_rx(struct usbh_cdc_ecm_data *const data);
+
+static void usbh_cdc_ecm_sig_raise(struct usbh_cdc_ecm_data *const data, unsigned int result)
+{
+	(void)atomic_or(&data->rx_pending_sig_vals, result);
+	(void)k_poll_signal_raise(data->rx_sig, 0);
+}
+
+static void usbh_cdc_ecm_start_auto_rx(struct usbh_cdc_ecm_data *const data)
+{
+	(void)atomic_set(&data->auto_rx_enabled, 1);
+
+	usbh_cdc_ecm_sig_raise(data, USBH_CDC_ECM_SIG_COMM_RX_IDLE | USBH_CDC_ECM_SIG_DATA_RX_IDLE);
+}
+
+static void usbh_cdc_ecm_stop_auto_rx(struct usbh_cdc_ecm_data *const data)
+{
+	(void)atomic_clear(&data->auto_rx_enabled);
+	(void)atomic_clear(&data->rx_pending_sig_vals);
+
+	k_poll_signal_reset(data->rx_sig);
+}
+
+static int usbh_cdc_ecm_comm_rx_cb(struct usb_device *const udev, struct uhc_transfer *const xfer)
+{
+	struct usbh_cdc_ecm_data *priv = xfer->priv;
+	struct cdc_notification_packet *notif;
+	uint32_t *link_speeds;
+	int ret = xfer->err;
+
+	if (atomic_get(&priv->auto_rx_enabled)) {
+		if (usbh_cdc_ecm_comm_rx(priv)) {
+			usbh_cdc_ecm_sig_raise(priv, USBH_CDC_ECM_SIG_COMM_RX_IDLE);
+		}
+	}
+
+	if (xfer->err) {
+		LOG_ERR("comm rx xfer callback error (%d)", ret);
+		goto cleanup;
+	}
+
+	notif = (struct cdc_notification_packet *)xfer->buf->data;
+	switch (notif->bNotification) {
+	case USB_CDC_NETWORK_CONNECTION:
+		if (xfer->buf->len != sizeof(struct cdc_notification_packet)) {
+			ret = -EBADMSG;
+			goto cleanup;
+		}
+		if (sys_le16_to_cpu(notif->wValue)) {
+			net_if_carrier_on(priv->iface);
+		} else {
+			usbh_cdc_ecm_stop_auto_rx(priv);
+			net_if_carrier_off(priv->iface);
+		}
+		break;
+	case USB_CDC_CONNECTION_SPEED_CHANGE:
+		if (xfer->buf->len != (sizeof(struct cdc_notification_packet) + 8)) {
+			ret = -EBADMSG;
+			goto cleanup;
+		}
+
+		link_speeds = (uint32_t *)(notif + 1);
+
+		for (size_t i = 0; i < 2; i++) {
+			link_speeds[i] = sys_le32_to_cpu(link_speeds[i]);
+			switch (link_speeds[i]) {
+			case 2500 * 1000000U:
+				priv->caps |= ETHERNET_LINK_2500BASE;
+			case 1000 * 1000000U:
+				priv->caps |= ETHERNET_LINK_1000BASE;
+			case 100 * 1000000U:
+				priv->caps |= ETHERNET_LINK_100BASE;
+			case 10 * 1000000U:
+				priv->caps |= ETHERNET_LINK_10BASE;
+				break;
+			default:
+				break;
+			}
+		}
+		break;
+	default:
+		break;
+	}
```

Comment:
the `caps` are just for the generally supported speeds of the device, not of the current speeds. if you want to expose this to the user, use the ethernet phy api.